ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java

```diff
@@ -0,0 +1,345 @@
+        } else {
+          codeTx.put(codeUpdate.getKey().toArrayUnsafe(), updatedCode.toArrayUnsafe());
+        }
+      }
+
+      // next collect the branches that will be trimmed
+      final StoredMerklePatriciaTrie<Bytes, Bytes> accountTrie =
+          new StoredMerklePatriciaTrie<>(
+              this::getTrieNode, worldStateRootHash, Function.identity(), Function.identity());
+
+      // for manicured tries and composting, collect branches here (not implemented)
+
+      // now add the accounts
+      for (final Map.Entry<Address, BonsaiValue<BonsaiAccount>> accountUpdate :
+          updater.getAccountsToUpdate().entrySet()) {
+        final Bytes accountKey = accountUpdate.getKey();
+        final BonsaiValue<BonsaiAccount> bonsaiValue = accountUpdate.getValue();
+        final BonsaiAccount updatedAccount = bonsaiValue.getUpdated();
+        if (updatedAccount == null) {
+          final Hash addressHash = Hash.hash(accountKey);
+          accountTrie.remove(addressHash);
+          accountTx.remove(accountKey.toArrayUnsafe());
+        } else {
+          final Hash addressHash = updatedAccount.getAddressHash();
+          final Bytes accountValue = updatedAccount.serializeAccount();
+          accountTx.put(accountKey.toArrayUnsafe(), accountValue.toArrayUnsafe());
+          accountTrie.put(addressHash, accountValue);
+        }
+      }
+
+      accountTrie.commit((location, hash, value) -> writeTrieNode(trieBranchTx, location, value));
+      worldStateRootHash = accountTrie.getRootHash();
+      trieBranchTx.put(WORLD_ROOT_KEY, worldStateRootHash.toArrayUnsafe());
+
+      // for manicured tries and composting, trim and compost branches here
+
+      if (blockHash != null) {
+        final TrieLogLayer trieLog = updater.generateTrieLog(blockHash);
+        trieLog.freeze();
+        // TODO add to archive here, but only once we get persisted follow distance implemented
+        // archive.addLayeredWorldState(new BonsaiLayeredWorldState(this, trieLog));
+
+        final BytesValueRLPOutput rlpLog = new BytesValueRLPOutput();
+        trieLog.writeTo(rlpLog);
+        trieLogTx.put(blockHash.toArrayUnsafe(), rlpLog.encoded().toArrayUnsafe());
+      }
+
+      success = true;
+    } finally {
+      if (success) {
+        accountTx.commit();
+        codeTx.commit();
+        storageTx.commit();
+        trieBranchTx.commit();
+        trieLogTx.commit();
+        updater.reset();
+      } else {
+        accountTx.rollback();
+        codeTx.rollback();
+        storageTx.rollback();
+        trieBranchTx.rollback();
+        trieLogTx.rollback();
+      }
+    }
+  }
+
+  private static Bytes rlpEncode(final Bytes bytes) {
+    final BytesValueRLPOutput out = new BytesValueRLPOutput();
+    out.writeBytes(bytes.trimLeadingZeros());
+    return out.encoded();
+  }
+
+  @Override
+  public WorldUpdater updater() {
+    if (updater == null) {
+      updater = new BonsaiWorldStateUpdater(this);
```

Comment:
As I get to the full lagging persisted state this will change.  Right now no layered world states sit between the updater and the persisted state, and there is only one persisted and one pending state.  Expect this to change dramatically.